### PR TITLE
feat(optimize): add keyword rank evaluation

### PR DESCRIPTION
**Pull request by rudrankriyam**
## Problem

Agents are good at brainstorming App Store keyword candidates and bad at knowing whether any of them are worth using. The repository already ships every ingredient needed to *evaluate* a candidate list — public storefront search, ranking, and Apple Ads demand data — but nothing composes them into a keyword workflow. This is the first PR of a three-PR stack that adds `asc optimize keywords`:

1. **this PR** — `rank` (zero-auth, public iTunes)
2. `score` — composed difficulty/popularity/competition scoring
3. `discover` — official Apple Ads suggestion endpoints

## Behavior

`asc optimize keywords rank` reports where an app sits in the public App Store search result window for each keyword candidate. No authentication is required.

```
asc optimize keywords rank --app APP_ID --keywords LIST [--country us] [--platform IOS|TV_OS] [--workers N]
```

Keywords are normalized before anything is requested: trimmed, lowercased, internal whitespace collapsed, and deduplicated while preserving caller order. Each invocation accepts at most 100 keywords of 2-60 characters and at most 4 space-separated words, and the usage error names the limit it hit.

Lookups fan out with bounded concurrency (`--workers`, default 10, minimum 1). A keyword whose lookup fails becomes an `unavailable` row carrying its error rather than being dropped or silently scored as absent; the command fails only when *every* keyword fails, and then reports one stable representative error (server failures outrank client failures, lowest status wins within a class) so the reported cause does not depend on goroutine completion order.

Row status reuses the official Apple Ads source-status vocabulary already in the tree (`available` / `empty` / `unavailable`) so an unavailable lookup is never confused with a genuine absence. `rank` serializes as an explicit `null` when the app is absent, and `totalResults` is omitted entirely for failed lookups rather than being reported as zero.

### Example

```console
$ asc optimize keywords rank --app 1234567890 --keywords "focus timer,habit tracker,broken keyword" --country us
```

```json
{
  "schemaVersion": "1",
  "generatedAt": "2026-08-19T05:12:44Z",
  "appId": "1234567890",
  "country": "US",
  "platform": "IOS",
  "workers": 10,
  "summary": { "keywords": 3, "ranked": 1, "absent": 1, "unavailable": 1 },
  "rows": [
    { "keyword": "focus timer",    "rank": 2,    "totalResults": 200, "status": "available" },
    { "keyword": "habit tracker",  "rank": null, "totalResults": 200, "status": "empty" },
    { "keyword": "broken keyword", "rank": null, "status": "unavailable", "error": "search request returned status 429" }
  ]
}
```

`--output table` and `--output markdown` render a summary section plus one row per keyword.

## Formula transparency

This command computes no score — it reports Apple's ordering directly. Every row ships the raw inputs behind it (`totalResults`, `status`, `error`), so the result is reproducible from the same public endpoint. The stack's stated design position, that unavailable official sources are reported rather than replaced with invented values, is preserved here: nothing is fabricated for a failed lookup. The scoring model arrives in PR 2 alongside `docs/design/optimize-keywords.md`.

## Compatibility

Additive only. A new `keywords` subgroup is registered inside the existing experimental `optimize` tree; no existing command, flag, or output shape changes. `docs/COMMANDS.md` lists top-level commands only, so it is unchanged (verified via `make generate-command-docs` + `make check-docs`).

The `optimize` tree marks stability with a trailing `... [experimental]` **suffix** (`internal/cli/optimize/command.go`), whereas `TestExperimentalCommandsHaveStabilityLabel` in `internal/cli/cmdtest/stability_tiers_test.go` asserts a **prefix** for the surfaces it enumerates. I matched the siblings in this tree deliberately rather than splitting the `optimize` group's own convention; the new commands are not in that test's enumeration. Worth a maintainer decision on which convention wins tree-wide — happy to flip it either way.

## Tests

- `internal/cli/optimize/keywords_test.go` — keyword hygiene (trim/lowercase/collapse/dedupe, boundary values, every rejection message), fan-out concurrency bounding and order preservation, failure accumulation without cancelling peers, representative-error selection table.
- `internal/cli/optimize/keywords_rank_test.go` — `httptest` coverage for rank found / absent / per-keyword failure, explicit `null` rank serialization, all-keywords-failed error path, `--workers` request bounding, table and markdown rendering, and every usage error asserted to run before any client is constructed.
- `internal/cli/cmdtest/optimize_keywords_rank_test.go` — end-to-end through the real root command: help wiring, usage errors on stderr with no stdout, full JSON contract, and the all-failed path.

Gauntlet: `make build`, `make format`, `make check-docs`, `make lint` (0 issues), `ASC_BYPASS_KEYCHAIN=1 make test` — all green, no pre-existing failures. No live API calls; all coverage is `httptest`/round-tripper fixtures.

## Summary by CodeRabbit

* **New Features**
  * Added an experimental `optimize keywords rank` command for checking App Store keyword rankings.
  * Supports normalized, deduplicated keywords; configurable app, country, and iOS or tvOS platform settings.
  * Reports ranked, absent, and unavailable keywords with summary counts and concise error details.
  * Provides table, Markdown, and JSON output formats.
  * Supports bounded concurrent lookups and cancellation handling.
* **Bug Fixes**
  * Improved validation for invalid, zero, and overflowing app IDs.
  * Handles partial and complete lookup failures with clear, deterministic errors.

### CODE REVIEW

File: internal/cli/optimize/keywords_rank.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+type KeywordRankSummary struct {
+	Keywords    int `json:"keywords"`
+	Ranked      int `json:"ranked"`
+	Absent      int `json:"absent"`
+	Unavailable int `json:"unavailable"`
+}
+
+// KeywordRankReport is the stable JSON contract emitted by
+// `asc optimize keywords rank`.
+type KeywordRankReport struct {
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_rank.go]: Fixed in 8516eb675. The computed report contract now lives in `internal/asc/output_optimize.go`; the command and renderers use those exported types, and a contract test verifies the camelCase JSON shape, including explicit `rank: null` and `totalResults` omission.

File: internal/cli/cmdtest/optimize_keywords_rank_test.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+			return jsonResponse(http.StatusOK, `{"resultCount":2,"results":[{"trackId":111},{"trackId":1234567890}]}`)
+		case "habit tracker":
+			return jsonResponse(http.StatusOK, `{"resultCount":1,"results":[{"trackId":111}]}`)
+		case "broken keyword":
+			return jsonResponse(http.StatusServiceUnavailable, `{}`)
+		default:
+			t.Fatalf("unexpected term %q", req.URL.Query().Get("term"))
+			return nil, errors.New("unexpected term")
+		}
+	})
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/cmdtest/optimize_keywords_rank_test.go:143]: Fixed in 3d0765c6d. Every transport callback failure now uses `t.Errorf` and returns an error to the worker, avoiding `t.Fatalf` from a non-test goroutine. The focused command tests and race-enabled end-to-end test pass.

**Comment by rudrankriyam** [comment]
@coderabbitai review

File: internal/cli/optimize/keywords.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	resolved := strings.TrimSpace(shared.ResolveAppID(appID))
+	if resolved == "" {
+		return "", shared.UsageError("--app is required (or set ASC_APP_ID)")
+	}
+	for _, digit := range resolved {
+		if digit < '0' || digit > '9' {
+			return "", shared.UsageError("--app must be a numeric App Store app ID")
+		}
+	}
+	return resolved, nil
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords.go:121]: Fixed in c06986024. The existing digits-only validation now also parses the ID as a signed 64-bit integer and requires it to be positive before constructing the client. Regression cases cover zero and one-past-MaxInt64 and assert that no request path is reached.

**Comment by rudrankriyam** [comment]
@coderabbitai review

File: internal/cli/optimize/keywords_rank.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			})
+			if report.Summary.Unavailable == report.Summary.Keywords {
+				failures := make([]error, 0, len(results))
+				for _, result := range results {
+					failures = append(failures, result.Err)
+				}
+				return fmt.Errorf("optimize keywords rank: %w", representativeKeywordError(failures))
+			}
+
+			return shared.PrintOutputWithRenderers(
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_rank.go]: Fixed in 5e830d837. `KeywordRankReport` now has a direct registry renderer owned beside its contract, with summary and keyword row extraction covered by a registry-level regression test. The command now passes `*KeywordRankReport` through `shared.PrintOutput`, and the obsolete command-local renderer file is removed.

**Comment by rudrankriyam** [comment]
@coderabbitai review

File: internal/cli/optimize/keywords_rank.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+			}
+
+			client := newKeywordsItunesClient()
+			results := fanOutKeywords(ctx, normalizedKeywords, *workers, func(ctx context.Context, keyword string) (itunes.PublicRankResult, error) {
+				requestCtx, cancel := shared.ContextWithTimeout(ctx)
+				defer cancel()
+				return client.RankApp(requestCtx, resolvedAppID, keyword, normalizedCountry, normalizedPlatform)
+			})
+
+			report := buildKeywordRankReport(keywordRankBuildInput{
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_rank.go:105]: Fixed in 921235c5c. The command now returns the parent context error immediately after fan-out and before report construction or rendering. A deterministic regression makes one lookup succeed, cancels during the second lookup, and verifies `context.Canceled` with no stdout; race coverage and the full suite pass.

**Comment by rudrankriyam** [comment]
@coderabbitai review
